Improve manage_raid resouce code and test

enrico-usai opened the pull request:
### Description of changes
* Move `mdadm` package installation into `manage_raid` resource
  * Define `:setup` as default action and execute it at install time.
* Rename `manage_raid` custom resource to `raid`
  * Now we have the `:setup` action so it is a more appropriate name.
* Rename head_node_raid recipe to raid_head_node
* Rename compute_raid to raid_compute

### Tests
* Add extra checks in the EBS tests
  * These checks are taken from integration tests (CLI) code.
* Define Inspec test for `mdadm` installation. This package is not installed by default on Centos7 so with the test we verified it is installed.
* Extend raid Inspec tests (test code taken from integration tests)
  * Ensure that the RAID array is reassembled automatically on boot
  * Verify RAID is correctly mounted
* `bash kitchen.ec2.sh environment-config test raid-mounted-centos7`
* `bash kitchen.docker.sh environment-install test raid-centos7`

### References
* https://github.com/aws/aws-parallelcluster/pull/5385

Commit:
Add extra checks in the EBS tests

These checks are taken from integration tests (CLI) code.

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Move mdadm package installation into raid resource

Define `:setup` as default action and execute it at install time.

### Tests

Define Inspec test for `mdadm` installation.
This package is not installed by default on Centos7
so with the test we verified it is installed.

Verified with:
* `bash kitchen.docker.sh environment-install test raid-centos7`

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Extend raid Inspec tests

* Verify RAID is correctly configured
* Ensure that the RAID array is reassembled automatically on boot
* Verify RAID is correctly mounted

Test code taken from integration tests.

### Tests
* `bash kitchen.ec2.sh environment-config test raid-mounted-centos7`

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Rename manage_raid custom resource to raid

Now we have a :setup action so raid is a more appropriate name

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Rename head_node_raid recipe to raid_head_node

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Rename compute_raid recipe to raid_compute

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Unify raid recipes

Code for head node and compute fleet and conditions are now into the same recipe.

Signed-off-by: Enrico Usai <usai@amazon.com>

awslitvit commented on cookbooks/aws-parallelcluster-environment/test/controls/ebs_spec.rb:
```diff
@@ -17,6 +17,18 @@
+
+  describe 'Verify EBS is correctly mounted'
+
+  describe bash("df -h -t ext4 | tail -n +2 | awk '{{print $2, $6}}' | grep '/shared_dir'") do
```
Nit: it's worth adding a comment to explain what this command does.

enrico-usai replied:
Added

awslitvit commented on cookbooks/aws-parallelcluster-environment/test/controls/ebs_spec.rb:
```diff
@@ -17,6 +17,18 @@
+
+  describe bash("cat /etc/fstab") do
+    its('exit_status') { should eq(0) }
+    its('stdout')      { should match('UUID=.* /shared_dir ext4 _netdev 0 0') }
```
Do we still need an integration test? Could we split this control and mark its part which does not depend on how the unit test is set up with `tag:config` so that it is checked daily as part of kitchen tests?

enrico-usai replied:
We still need integration tests because we're testing the different flavours of raid (0 and 1) and with multiple devices (5). In future we can extend kitchen tests and remove

enrico-usai replied:
Added

awslitvit commented on cookbooks/aws-parallelcluster-environment/test/controls/raid_spec.rb:
```diff
@@ -17,6 +23,33 @@
+
+  describe 'Verify RAID is correctly mounted'
+
+  describe bash("df -h -t ext4 | tail -n +2 | awk '{{print $2, $6}}' | grep '/shared_dir'") do
```
Same as above.

enrico-usai replied:
Added

awslitvit commented on cookbooks/aws-parallelcluster-platform/resources/install_packages/install_packages_amazon2.rb:
```diff
@@ -26,7 +26,7 @@ def default_packages
      libXmu-devel hwloc-devel libdb-devel tcl-devel automake autoconf pyparted libtool
      httpd boost-devel system-lsb mlocate atlas-devel glibc-static iproute
      libffi-devel dkms libedit-devel postgresql-devel postgresql-server
-     sendmail cmake byacc libglvnd-devel mdadm libgcrypt-devel libevent-devel
```
Love it!

awslitvit commented on cookbooks/aws-parallelcluster-environment/recipes/config/raid.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+when 'ComputeFleet'
+  unless raid_shared_dir.empty?
```
Why is this checked only for the compute fleet and not for the head node?

enrico-usai replied:
Code improved, using volume resource

Commit:
Put in common raid_shared_dir check

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Use volume resource for raid nfs mount steps

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Skip mdadm installation on Rhel8 on docker

Tested on ec2 and it passes.
On docker it fails with "No candidate package available"

Signed-off-by: Enrico Usai <usai@amazon.com>

Commit:
Add comments in the Inspec tests

Signed-off-by: Enrico Usai <usai@amazon.com>

awslitvit commented on cookbooks/aws-parallelcluster-environment/recipes/config/raid.rb:
```diff
@@ -1,7 +1,7 @@
 
 #
-# Copyright:: 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# Copyright:: 2013-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Nit: 2023

enrico-usai replied:
fixed
